### PR TITLE
[color] Use graphemes to measure strings.

### PR DESCRIPTION
**Pull request by pgavlin**
The number of Unicode code points in a string is not the same as the
number of user-visible characters (graphemes). When measuring colorized
strings, we want the latter rather than the former. Notably, these
changes fix some issues where the interactive display cut off before the
right edge of the terminal.

### CODE REVIEW

**Comment by pgavlin** [comment]
Stacks on top of https://github.com/pulumi/pulumi/pull/11201

File: pkg/backend/display/testdata/not-truncated/template-body.json.interactive-80x24-cooked.stdout.txt [hunk 1]
```diff
@@ -221,2303 +221,2303 @@
...
- <{%reset%}>  <{%reset%}>     ├─ eks:index:ServiceRole       cluster-eksRole                  <{%reset%}><{%reset%}>   
- <{%reset%}>  <{%reset%}>     │  └─ aws:iam:Role             cluster-eksRole-role             <{%reset%}><{%reset%}> 
- <{%reset%}>  <{%reset%}>     ├─ eks:index:RandomSuffix      cluster-cfnStackName             <{%reset%}><{%reset%}>   
- <{%bold%}><{%reset%}>  <{%reset%}>     └─ aws:ec2:SecurityGroup       cluster-eksClusterSecurityGroup  <{%bold%}><{%reset%}><{%reset%}>   
+ <{%reset%}>  <{%reset%}>  ├─ awsx:x:ec2:Vpc                 vpc                              <{%reset%}><{%reset%}>       
+ <{%reset%}>  <{%reset%}>  │  ├─ awsx:x:ec2:NatGateway       vpc-0                            <{%reset%}><{%reset%}>       
+ <{%reset%}>  <{%reset%}>  │  │  └─ aws:ec2:Eip              vpc-0                            <{%reset%}><{%reset%}>       
+ <{%reset%}>  <{%reset%}>  │  ├─ awsx:x:ec2:Subnet           vpc-public-0                     <{%reset%}><{%reset%}>       
+ <{%reset%}>  <{%reset%}>  │  │  └─ aws:ec2:RouteTable       vpc-public-0                     <{%reset%}><{%reset%}>
```
Comment (AaronFriel) [inline pkg/backend/display/testdata/not-truncated/template-body.json.interactive-80x24-cooked.stdout.txt:651]: Nit: It looks like we can optimize out a colorized column of length 0 and elide it.

**Comment by pgavlin** [comment]
> Notably, these changes fix some issues where the interactive display cut off before the right edge of the terminal.

To elaborate on these issues:

In the interactive display, resources are rendered in a tree view with visible edges from children to parents. Each of these edges is formed using Unicode Box Drawing characters. These characters are outside the ASCII range, and each of their UTF-8 code units consists of three bytes. The code in the colorizer that clamps the output of colorization to a particular width was using byte counts instead of code points or graphemes. With these changes, these measurements are now done using graphemes (i.e. user-visible characters).

File: pkg/backend/display/testdata/not-truncated/template-body.json.interactive-100x80-cooked.stdout.txt [hunk 2]
```diff
@@ -2770,1230 +2770,1230 @@
...
+ <{%reset%}>  <{%reset%}>     ├─ aws:ec2:SecurityGroupRule         cluster-eksClusterIngressRule              <{%reset%}><{%reset%}>           
+ <{%reset%}>  <{%reset%}>     ├─ aws:ec2:SecurityGroupRule         cluster-eksNodeInternetEgressRule          <{%reset%}><{%reset%}>           
+ <{%reset%}>  <{%reset%}>     ├─ aws:ec2:SecurityGroupRule         cluster-eksNodeIngressRule                 <{%reset%}><{%reset%}>           
+ <{%reset%}>  <{%reset%}>     ├─ aws:ec2:SecurityGroupRule         cluster-eksNodeClusterIngressRule          <{%reset%}><{%reset%}>           
+ <{%reset%}>  <{%reset%}>     ├─ kubernetes:core/v1:ConfigMap      cluster-nodeAccess                         <{%reset%}><{%reset%}>           
+ <{%reset%}>  <{%reset%}>     ├─ aws:ec2:LaunchConfiguration       cluster-nodeLaunchConfiguration            <{%reset%}><{%reset%}>           
+ <{%bold%}><{%fg 3%}>~ <{%reset%}>     └─ aws:cloudformation:Stack          cluster-nodes                              <{%bold%}><{%fg 3%}>updating<{%reset%}><{%bold%}><{%fg 3%}><{%reset%}>
```
Comment (AaronFriel) [inline pkg/backend/display/testdata/not-truncated/template-body.json.interactive-100x80-cooked.stdout.txt:3584]: Kudos: Having snapshot based tests for display logic make reviewing these PRs much easier. 👍

**Review by AaronFriel — APPROVED** [review]
Oops. Meant to approve. Nit is non-blocking, per name.

**Comment by pgavlin** [comment]
bors merge

**Comment by pgavlin** [comment]
bors retry

**Pull request merged**